spec-cordova/prepare.spec.js

```diff
@@ -17,175 +17,265 @@
+            spyOn(cordova_config, 'read').and.returnValue({});
+            spyOn(restore, 'installPlatformsFromConfigXML').and.returnValue(Q());
+            spyOn(restore, 'installPluginsFromConfigXML').and.returnValue(Q());
+            
+            spyOn(prepare, 'preparePlatforms').and.returnValue(Q);
+        });
+        describe('failure', function () {
+            it('should invoke util.preProcessOptions as preflight task checker, which, if fails, should trigger promise rejection and only fire the before_prepare hook', function(done) {
+                spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
+                spyOn(util, 'preProcessOptions').and.throwError();
+                prepare({}).fail(function(err) {
+                    expect(err).toBeDefined();
+                    expect(HooksRunner.prototype.fire.calls.count()).toBe(1);
+                    expect(HooksRunner.prototype.fire.calls.argsFor(0)[0]).toEqual('before_prepare');
+                    done();
+                });
+            });
+            it('should invoke util.cdProjectRoot as a preflight task checker, which, if fails, should trigger a promise rejection and fire no hooks', function(done) {
+                spyOn(util, 'cdProjectRoot').and.throwError();
+                prepare({}).fail(function(err) {
+                    expect(err).toBeDefined();
+                    expect(HooksRunner.prototype.fire.calls.any()).toBe(false);
+                    done();
+                });
+            });
         });
-        it('Test 002 : should not run inside a cordova-based project with no platforms', function(done) {
-            list_platforms.and.returnValue([]);
-            prepare().then(function() {
-                expect('this call').toBe('fail');
-            }, function(err) {
-                expect('' + err).toContain('No platforms added to this project. Please use `cordova platform add <platform>`.');
-            }).fin(done);
+
+        describe('success', function () {
+            beforeEach(function () {
+                spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
+                spyOn(util, 'preProcessOptions').and.callFake(function(options) {
+                    let platforms = options.platforms || [];
+                    return {'platforms':platforms};
+                });
+            });
+            it('should fire the before_prepare hook and provide platform and path information as arguments', function(done) {
+                prepare({}).then(function() {
+                    expect(HooksRunner.prototype.fire.calls.argsFor(0)[0]).toEqual('before_prepare');
+                    done();
+                });
+            });
+            it('should invoke restore module\'s installPlatformsFromConfigXML method', function(done) {
+                prepare({}).then(function() {
+                    expect(restore.installPlatformsFromConfigXML).toHaveBeenCalled();
+                    done();
```

Comment:
Instead of popping a `done()` into both the success and error handlers, factor it out into a final `.done(done);` promise handler.